core/src/main/java/com/linecorp/armeria/internal/eventloop/EventLoopCheckingCompletableFuture.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2019 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.internal.eventloop;
+
+import java.util.Collections;
+import java.util.Set;
+import java.util.WeakHashMap;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.linecorp.armeria.common.Flags;
+
+import reactor.core.scheduler.NonBlocking;
+
+/**
+ * A {@link CompletableFuture} that warns the user if they call a method that blocks the event loop.
+ */
+public final class EventLoopCheckingCompletableFuture<T> extends CompletableFuture<T> {
+
+    private static final Logger logger = LoggerFactory.getLogger(EventLoopCheckingCompletableFuture.class);
+
+    private static final Set<Thread> REPORTED_THREADS = Collections.synchronizedSet(
+            Collections.newSetFromMap(new WeakHashMap<>()));
+
+    @Override
+    public T get() throws InterruptedException, ExecutionException {
+        maybeLogIfOnEventLoop();
+        return super.get();
+    }
+
+    @Override
+    public T get(long timeout, TimeUnit unit)
+            throws InterruptedException, ExecutionException, TimeoutException {
+        maybeLogIfOnEventLoop();
+        return super.get(timeout, unit);
+    }
+
+    @Override
+    public T join() {
+        maybeLogIfOnEventLoop();
+        return super.join();
+    }
+
+    private static void maybeLogIfOnEventLoop() {
+        if (!Flags.reportBlockedEventLoop()) {
+            return;
+        }
+        final Thread thread = Thread.currentThread();
+        if (thread instanceof NonBlocking && REPORTED_THREADS.add(thread)) {
```

Comment:
IIUC this [`NonBlocking` interface](https://github.com/reactor/reactor-core/blob/master/reactor-core/src/main/java/reactor/core/scheduler/NonBlocking.java) is created in the project reactor.

https://github.com/reactor/reactor-core/blob/628258f644bdddd095e72d1fc81c5d6ac0e3fa12/reactor-core/src/main/java/reactor/core/scheduler/ReactorThreadFactory.java#L92-L96

It seems an edge case but a [non-eventloop thread](https://github.com/reactor/reactor-core/blob/628258f644bdddd095e72d1fc81c5d6ac0e3fa12/reactor-core/src/main/java/reactor/core/scheduler/ReactorThreadFactory.java#L92) might pass this condition and log this message. I am wondering if it is acceptable. Unless I suggest changing `EventLoopCheckingCompletableFuture` to `NonBlockCheckingCompletableFuture` and related message. 😀

Reply:
Armeria `EventLoopThread` implements `Nonblocking`. 😄

Reply:
I originally checked `instanceof EventLoopThread` but changed to `NonBlocking` I think to avoid moving a package-local `EventLoopThread` to public in `internal` package. @trustin Any thoughts? I lean towards it being small enough not to worry about

Reply:
I know that :-) Just worried other non-event loop thread could implement `NonBlocking`. 
But it seems too many worries. 😜

Reply:
I think we can warn that as well because a thread which implements `NonBlocking` is doing blocking operation. 😄

Reply:
Originally I wasn't worried about the behavior. My concern is the log message.
Sorry for the confusion. 😅
```diff
- Calling a blocking method on CompletableFuture from an event loop thread.
+ Calling a blocking method on CompletableFuture from a non-blocking thread.
```

Reply:
> I lean towards it being small enough not to worry about

But I agree with this opinion!

Reply:
Ah then it makes sense. 😄

Reply:
So do we need to update the WARN log message?

Reply:
I'm not completely sure it is worth to update the message.
If we do, then leave the original sentence and put extra information just in case. 😀

```diff
- Calling a blocking method on CompletableFuture from an event loop thread.
+ Calling a blocking method on CompletableFuture from an event loop thread or non-blocking thread.
```

Reply:
@anuraaga Could you update the log message?

Reply:
Sure